STYLE.md

```diff
@@ -0,0 +1,4068 @@
+  precomputation.
+
+### Exit in Main
+
+Go programs use [`os.Exit`](https://pkg.go.dev/os#Exit) or [`log.Fatal*`](https://pkg.go.dev/log#Fatal) to exit immediately. (Panicking
+is not a good way to exit programs, please [don't panic](#dont-panic).)
+
+Call one of `os.Exit` or `log.Fatal*` **only in `main()`**. All other
+functions should return errors to signal failure.
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+func main() {
+  body := readFile(path)
+  fmt.Println(body)
+}
+
+func readFile(path string) string {
+  f, err := os.Open(path)
+  if err != nil {
+    log.Fatal(err)
+  }
+
+  b, err := io.ReadAll(f)
+  if err != nil {
+    log.Fatal(err)
+  }
+
+  return string(b)
+}
+```
+
+</td><td>
+
+```go
+func main() {
+  body, err := readFile(path)
+  if err != nil {
+    log.Fatal(err)
+  }
+  fmt.Println(body)
+}
+
+func readFile(path string) (string, error) {
+  f, err := os.Open(path)
+  if err != nil {
+    return "", err
+  }
+
+  b, err := io.ReadAll(f)
+  if err != nil {
+    return "", err
+  }
+
+  return string(b), nil
+}
+```
+
+</td></tr>
+</tbody></table>
+
+Rationale: Programs with multiple functions that exit present a few issues:
+
+- Non-obvious control flow: Any function can exit the program so it becomes
+  difficult to reason about the control flow.
+- Difficult to test: A function that exits the program will also exit the test
+  calling it. This makes the function difficult to test and introduces risk of
+  skipping other tests that have not yet been run by `go test`.
+- Skipped cleanup: When a function exits the program, it skips function calls
+  enqueued with `defer` statements. This adds risk of skipping important
+  cleanup tasks.
+
+#### Exit Once
+
+If possible, prefer to call `os.Exit` or `log.Fatal` **at most once** in your
+`main()`. If there are multiple error scenarios that halt program execution,
+put that logic under a separate function and return errors from it.
+
+This has the effect of shortening your `main()` function and putting all key
+business logic into a separate, testable function.
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+package main
+
+func main() {
+  args := os.Args[1:]
+  if len(args) != 1 {
+    log.Fatal("missing file")
+  }
+  name := args[0]
+
+  f, err := os.Open(name)
+  if err != nil {
+    log.Fatal(err)
+  }
+  defer f.Close()
+
+  // If we call log.Fatal after this line,
+  // f.Close will not be called.
+
+  b, err := io.ReadAll(f)
+  if err != nil {
+    log.Fatal(err)
+  }
+
+  // ...
+}
+```
+
+</td><td>
+
+```go
+package main
+
+func main() {
+  if err := run(); err != nil {
+    log.Fatal(err)
+  }
+}
+
+func run() error {
+  args := os.Args[1:]
+  if len(args) != 1 {
+    return errors.New("missing file")
+  }
+  name := args[0]
+
+  f, err := os.Open(name)
+  if err != nil {
+    return err
+  }
+  defer f.Close()
```

Comment:
Note that this swallows error from `f.Close()` which is not good.
There's a pattern that helps with error handling in `defer`:  https://wstrm.dev/posts/errors-join-heart-defer/
Discussion on `/r/golang`: https://www.reddit.com/r/golang/comments/11luvxm/errorsjoin_defer/
Also: https://stackoverflow.com/a/76287159
IMO it has a downside, namely that you must not forget to use named return value (which by itself is sometimes frowned upon), but still it's probably better than alternatives

```go
func run() (err error) {
  args := os.Args[1:]
  if len(args) != 1 {
    return errors.New("missing file")
  }
  name := args[0]

  f, err := os.Open(name)
  if err != nil {
    return err
  }
  defer func () {
    err = errors.Join(err, f.Close())
  }()

  b, err := io.ReadAll(f)
  if err != nil {
    return err
  }

  // ...
}
```

If we're to use this pattern, it probably belongs to the section on `defer`

Reply:
I'm not sure if I like this pattern. It doesn't handle the error. If closing fails this should in my opinion be handled where the closing happened and not further up the call stack. This pattern only achieves that the error isn't lost.

Also this introduces possible false positives:

```go
func run() (err error) {
	f, _ := os.Create("test.txt")
	defer func() {
		err = errors.Join(err, f.Close())
	}()

	f.Write([]byte("test"))
	if err := f.Close(); err != nil { // closed explicitly to handle error
		// handle err instead of just logging it
		log.Fatal(err)
	}
	return nil
}

func main() {
	fmt.Println(run())
}
```

This prints the error "file closed already" which isn't an error and should be filtered, making the `defer`red function unnecessarily complex 😕

Reply:
The problem is that if we have many clauses like this in `run()`
```go
  b, err := io.ReadAll(f)
  if err != nil {
    return err
  }

  if err := doSmthWithBuffer(b); err != nil {
    return fmt.Errorf("smth: %w", err)
  }

  if err := doSmthElse(b); err != nil {
    return fmt.Errorf("smth else: %w", err)
  }
```
we would have to rewrite them like
```go
```
  b, err := io.ReadAll(f)
  if err != nil {
    return errors.Join(err, f.Close())
  }

  if err := doSmthWithBuffer(b); err != nil {
    return errors.Join(fmt.Errorf("smth: %w", err), f.Close())
  }

  if err := doSmthElse(b); err != nil {
    return errors.Join(fmt.Errorf("smth else: %w", err), f.Close())
  }
```
which quickly becomes error-prone, and also it's easy to shoot yourself in the foot by missing `f.Close()` in one of the branches as the code evolves. Moreover, if instead of just closing the file we had some more advanced cleanup that must run even if the code panics somewhere, it will fail to run.
Basically that's the situation for which `defer` exists, but we also need to handle errors that happen due to cleanup. Good if we can just log them and that's it, but such an approach is not always applicable.

Of course care must be taken not to close the file twice and not to handle the error twice (log and return), so in some cases the `defer` approach is not appropriate. But if you need to always close the file at the end of the function, this works quite well as you don't miss any errors and still have benefits of `defer`.

Reply:
But you are not handling the error? you are just appending it to another error - it might even be caused by the first error to begin with?

Instead of an anonymous deferred function the closing should be handled properly:

```go
func closeWithErrorHandling(f io.Closer) {
    err := f.close()
    if err == nil {
        return
    }
    // handle error
}

func run() error {
    f, _ := os.Create("test.txt")
    defer closeWithErrorHandling(f)

    // proceed doing with `f` what needs to be done
}

Reply:
This kind of error handling will not cause `run()` to return an error, and oftentimes we might want to indicate that `run` has failed to the caller instead of (or in addition to) doing any special handling in the `run`.
Of course if `run` can do all the handling we need and then return `nil` error such as log the error this pattern is unnecessary, but that is not what we always need.

So, just to be clear, the circumstances that warrant the use of the pattern are:
* a function needs to perform some cleanup when returning
* the cleanup may fail
* there are multiple `return` points in the function upon which the cleanup must be performed
* the function cannot perform all the necessary error handling on its own and must return an error if cleanup fails

This may sound a bit too specific but it's actually rather common situation

Reply:
but run will return an error if it's a different one from the one caused by `Close`? The error from close doesn't need to be returned it has already been handled...

Reply:
Whether handling errors in `defer` in case of failure is counterproductive is debatable, depending on the situation at hand.
E.g. if we're closing the database after a failed query, normally closing the database will succeed if the problem is a simple SQL error, but if the problem has happened due to some severe database corruption, failed `Close` may give some additional hints. IoW: in the majority of cases, the cleanup should succeed even on the "unhappy" path, and in rare cases where it fails the cleanup failure may provide additional hints.

Besides, a function may have several happy paths with early returns in addition to error returns, and upon such "happy" returns we may want to perform the cleanup and let the caller know if smth went wrong even  at that phase.

Reply:
I'd like to chip in here because this showed up in your PR @ivan4th and I also commented on this.

I am at the opinion that if we have to add this sort of cleanup code that has all sorts of complicated error instrumentation that is just difficult to understand and is not so informative (how do you know which line in the error corresponds to which `errors.Join`, etc), it is probably cleaner and nicer to do the cleanup _once_ on the calling context (i.e. the context that called the function that has to do the convoluted cleanup logic). Having these sort of anonymous functions is really difficult to read (to whoever didn't write them usually).

As for `errors.Join`, I find it useful in a case of a task that is parallelized (like in an `errgroup`, but then again `errgroup` should do this for you anyway a-la-carte).

On a personal note: I've seen maybe a case or two  where this sort of cleanup logic was needed so far (not in this project) and it was in a p2p protocol, where developer error was very easy to do and have caused certain cases where a `return` statement could miss satisfying all sorts of other calls that had to be done (cleanup of the libp2p stream, logging, metrics, etc), which is why it was later decided to have an anonymous function that catches whether an error was caught, causing the cleanup to be made more nicely (however, the code-smell is that the protocol was already way to convoluted, and this is why this was needed in the first place).

My suggestion here is that maybe we could also do a pair/mob programming session on the code in question (in the sql vaacum PR) so that we could see together whether the `errors.Join` is needed or not.

Reply:
I'm not insisting on the use of said pattern, if others don't like it, I will not use it, and I will update vacuum PR accordingly :)
If smth like that is to be used, simple `errors.Join` is probably indeed not the best approach, a cleanup errors need to be augmented, too, to make it cleaner what really happened. 

Moving cleanup to the calling context may often make sense, but in some cases it may also cause proliferation of funcs like `doSomethingInner`, `doInnerSomethingOfAnotherInnerSomething`, etc, if the only purpose of extracting a part of function is making cleanup easier. 

But in any case, simple `defer f.Close()` in the original example can be said to constitute a bug, as it may swallow a real I/O error which happens when the file is closed.

Reply:
Re pair/mob programming session, that might be a good idea, too

Reply:
> But in any case, simple defer f.Close() in the original example can be said to constitute a bug, as it may swallow a real I/O error which happens when the file is closed.

Yes this is true but if that happens, it doesn't really matter that `f.Cose()` fails:

```go
f, err := os.Open("path/to/file")
if err != nil {
    return err
}
defer f.Close()
buf, err := io.ReadAll(f)
if err != nil {
    return fmt.Error("read file: %w", err) // if close fails it doesn't really matter; read already failed, close will fail for the same cause if it does
}
// do something with `buf`
// happy case reached we now close explicitly
if err := f.Close(); err != nil {
    return fmt.Errorf("closing file: %w", err)
}
return nil
```

Reply:
@fasmat my point is, the original code doesn't have any handling for `f.Close()` errors at all:
```go

  f, err := os.Open(name)
  if err != nil {
    log.Fatal(err)
  }
  defer f.Close()

  // If we call log.Fatal after this line,
  // f.Close will not be called.

  b, err := io.ReadAll(f)
  if err != nil {
    log.Fatal(err)
  }
```
Granted, we're reading the file not writing it, which kind of makes implications less severe, but still completely ignore `Close` errors is not correct. In this case, the `Close` error will be ignored even if `io.ReadAll` succeeds.

Reply:
So I think that @fasmat is arguing here that handling the `Close` doesn't really matter because probably the error is so severe that the application will exit anyway. From my personal experience, having a `db.Close` on some part of a component that is going to raise and exception anyway is _best-effort_ at best. If you can't `Close` a db it usually means that:
- you have some very severe I/O errors on the hardware level
- disk is out of space and you can't flush stuff to it anymore

In our problem domain, I believe that other issues are probable but I'm more familiar with the ones above. Mostly, in those cases you assume that the node is no longer able to remain consistent (once you're out of space you're screwed because you don't know whether a "large" write is successful vs "small" write which largely leaves you with the assumption of no-longer-consistent).

In those cases node-operator is needed to recover the node and usually you need to start over. Hence I support that `Close` is a best-effort and the error can be ignored. 

https://stackoverflow.com/questions/37977714/can-i-get-an-error-when-closing-a-file